Video Viewer: Exiting pip puts the video into inline and shrinks video content

danae404 opened the pull request:
#### 036ff9d5b8e45f0f9bfd99ad73c7adb791c65df6
<pre>
Video Viewer: Exiting pip puts the video into inline and shrinks video content
<a href="https://bugs.webkit.org/show_bug.cgi?id=277466">https://bugs.webkit.org/show_bug.cgi?id=277466</a>
<a href="https://rdar.apple.com/132932256">rdar://132932256</a>

Reviewed by Jer Noble.

This patch adds new WKWebView SPI _enterInWindow and
_exitInWindow. _toggleInWindow enters in-window if the current mode is
anything other than in-window, and exits in-window if the current mode is
in-window. This patch separates that logic into two different methods. It will
be followed by a Safari patch utilizing the new SPI.

This bug occurs when a window occludes a video that is in viewer / in-window mode,
during which Safari puts the video into PiP mode. Then, if the user clicks
the return button on the pip window, the video enters inline mode instead of
in window mode, and the layout doesn&apos;t happen successfully in some way.

This bug is caused by WebKit and Safari racing to both change the video
mode from PiP to in-window. When the return from pip button is clicked,
this triggers webkit to set the presentation mode to in-window.
But during the exit pip process, the window becomes unoccluded,
and Safari calls _toggleInWindow in response. However because webkit has already
set the mode to in-window, calling _toggleInWindow makes the video go into inline.

This patch would give Safari the control to call either _enterInWindow or
_exitInWindow when it knows which one it wants to do. Calling _enterInWindow when the
mode is already set to in-window, or _exitInWindow when it is not in
in-window, has no effect.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::enterInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::exitInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindowFullscreen): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.h:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.mm:
(WebCore::PlaybackSessionInterfaceMac::enterInWindowFullscreen):
(WebCore::PlaybackSessionInterfaceMac::exitInWindowFullscreen):
(WebCore::PlaybackSessionInterfaceMac::toggleInWindowFullscreen): Deleted.
* Source/WebCore/platform/mac/WebPlaybackControlsManager.h:
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:
(-[WebPlaybackControlsManager enterInWindow]):
(-[WebPlaybackControlsManager exitInWindow]):
(-[WebPlaybackControlsManager toggleInWindow]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _enterInWindow]):
(-[WKWebView _exitInWindow]):
(-[WKWebView _toggleInWindow]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::enterInWindowFullscreen):
(WebKit::PlaybackSessionModelContext::exitInWindowFullscreen):
(WebKit::PlaybackSessionManagerProxy::enterInWindow):
(WebKit::PlaybackSessionManagerProxy::exitInWindow):
(WebKit::PlaybackSessionModelContext::toggleInWindowFullscreen): Deleted.
(WebKit::PlaybackSessionManagerProxy::toggleInWindow): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::enterInWindowFullscreen):
(WebKit::WebViewImpl::exitInWindowFullscreen):
(WebKit::WebViewImpl::toggleInWindowFullscreen): Deleted.
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::enterInWindow):
(WebKit::PlaybackSessionManager::exitInWindow):
(WebKit::PlaybackSessionManager::toggleInWindow): Deleted.
* Tools/TestWebKitAPI/Tests/mac/InWindowFullscreen.mm:
(TestWebKitAPI::TEST(InWindowFullscreen, EnterAndExitChangesIsActive)):
(TestWebKitAPI::TEST(InWindowFullscreen, EnterChangesIsActiveWithoutUserGesture)):
(TestWebKitAPI::TEST(InWindowFullscreen, ToggleChangesIsActive)): Deleted.
(TestWebKitAPI::TEST(InWindowFullscreen, ToggleChangesIsActiveWithoutUserGesture)): Deleted.

Canonical link: <a href="https://commits.webkit.org/281703@main">https://commits.webkit.org/281703@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70ef4a555de7f04b771f7edcba6ba0d67f4c7b2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64637 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11253 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62836 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47741 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11528 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49084 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7802 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62740 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37307 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52585 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29913 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33996 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9820 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10166 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55842 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66366 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4650 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9949 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56451 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4671 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52562 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56642 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13537 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3846 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35870 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36952 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38045 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/mac/WebPlaybackControlsManager.h:
```diff
@@ -74,7 +74,8 @@ WEBCORE_EXPORT
 @property (getter=isPictureInPictureActive) BOOL pictureInPictureActive;
 @property BOOL canTogglePictureInPicture;
 - (void)togglePictureInPicture;
-- (void)toggleInWindow;
```
Rather than remove this declaration, I'd recommend leaving it in place and implementing it in terms of the new enter/exit methods.

jernoble reviewed (approved):
r=me with nit:

danae404 committed:
Video Viewer: Exiting pip puts the video into inline and shrinks video content
https://bugs.webkit.org/show_bug.cgi?id=277466
rdar://132932256

Reviewed by Jer Noble.

This patch adds new WKWebView SPI _enterInWindow and
_exitInWindow. _toggleInWindow enters in-window if the current mode is
anything other than in-window, and exits in-window if the current mode is
in-window. This patch separates that logic into two different methods. It will
be followed by a Safari patch utilizing the new SPI.

This bug occurs when a window occludes a video that is in viewer / in-window mode,
during which Safari puts the video into PiP mode. Then, if the user clicks
the return button on the pip window, the video enters inline mode instead of
in window mode, and the layout doesn't happen successfully in some way.

This bug is caused by WebKit and Safari racing to both change the video
mode from PiP to in-window. When the return from pip button is clicked,
this triggers webkit to set the presentation mode to in-window.
But during the exit pip process, the window becomes unoccluded,
and Safari calls _toggleInWindow in response. However because webkit has already
set the mode to in-window, calling _toggleInWindow makes the video go into inline.

This patch would give Safari the control to call either _enterInWindow or
_exitInWindow when it knows which one it wants to do. Calling _enterInWindow when the
mode is already set to in-window, or _exitInWindow when it is not in
in-window, has no effect.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::enterInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::exitInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindowFullscreen): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.h:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.mm:
(WebCore::PlaybackSessionInterfaceMac::enterInWindowFullscreen):
(WebCore::PlaybackSessionInterfaceMac::exitInWindowFullscreen):
(WebCore::PlaybackSessionInterfaceMac::toggleInWindowFullscreen): Deleted.
* Source/WebCore/platform/mac/WebPlaybackControlsManager.h:
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:
(-[WebPlaybackControlsManager enterInWindow]):
(-[WebPlaybackControlsManager exitInWindow]):
(-[WebPlaybackControlsManager toggleInWindow]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _enterInWindow]):
(-[WKWebView _exitInWindow]):
(-[WKWebView _toggleInWindow]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::enterInWindowFullscreen):
(WebKit::PlaybackSessionModelContext::exitInWindowFullscreen):
(WebKit::PlaybackSessionManagerProxy::enterInWindow):
(WebKit::PlaybackSessionManagerProxy::exitInWindow):
(WebKit::PlaybackSessionModelContext::toggleInWindowFullscreen): Deleted.
(WebKit::PlaybackSessionManagerProxy::toggleInWindow): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::enterInWindowFullscreen):
(WebKit::WebViewImpl::exitInWindowFullscreen):
(WebKit::WebViewImpl::toggleInWindowFullscreen): Deleted.
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::enterInWindow):
(WebKit::PlaybackSessionManager::exitInWindow):
(WebKit::PlaybackSessionManager::toggleInWindow): Deleted.
* Tools/TestWebKitAPI/Tests/mac/InWindowFullscreen.mm:
(TestWebKitAPI::TEST(InWindowFullscreen, EnterAndExitChangesIsActive)):
(TestWebKitAPI::TEST(InWindowFullscreen, EnterChangesIsActiveWithoutUserGesture)):
(TestWebKitAPI::TEST(InWindowFullscreen, ToggleChangesIsActive)): Deleted.
(TestWebKitAPI::TEST(InWindowFullscreen, ToggleChangesIsActiveWithoutUserGesture)): Deleted.

Canonical link: https://commits.webkit.org/281703@main
